<table style="border: none" align="left">
   <tr style="border: none">
      <th style="border: none"><font face="verdana" size="5" color="black"><b>Predicting customer churn with IBM Watson Machine Learning</b></th>
      <th style="border: none"><img src="https://github.com/pmservice/customer-satisfaction-prediction/blob/master/app/static/images/ml_icon_gray.png?raw=true" alt="Watson Machine Learning icon" height="40" width="40"></th>
  <tr style="border: none">
       <th style="border: none"><img src="https://github.com/pmservice/wml-sample-models/blob/master/spark/customer-satisfaction-prediction/images/users_banner_2-03.png?raw=true" width="600" alt="Icon"> </th>
   </tr>
</table>

This notebook contains steps and code to develop a predictive model, and start scoring new data. This notebook introduces commands for getting data and for basic data cleaning and exploration, pipeline creation, model training, model persistance to Watson Machine Learning repository, model deployment, and scoring.

Some familiarity with Python is helpful. This notebook uses Python 2.7 and Apache® Spark 2.0.

You will use a data set, **Telco Customer Churn**, which details anonymous customers data of telecommunication company. Use the details of this data set to predict customer churn which is very critical to business as it's easier to retain existing customers rather than acquiring new ones.

## Learning goals

The learning goals of this notebook are:

-  Load a CSV file into an Apache® Spark DataFrame.
-  Explore data.
-  Prepare data for training and evaluation.
-  Create an Apache® Spark machine learning pipeline.
-  Train and evaluate a model.
-  Persist a pipeline and model in Watson Machine Learning repository.
-  Explore and visualize prediction result using the plotly package.
-  Deploy a model for batch scoring using Wastson Machine Learning API.


## Contents

This notebook contains the following parts:

1.	[Setup](#setup)
2.	[Load and explore data](#load)
3.	[Create spark ml model](#model)
4.	[Persist model](#persistence)
5.	[Predict locally and visualize](#visualization)
6.	[Deploy and score in a Cloud](#scoring)
7.	[Summary and next steps](#summary)

<a id="setup"></a>
## 1. Setup

Before you use the sample code in this notebook, you must perform the following setup tasks:

-  Create a [Watson Machine Learning Service](https://console.ng.bluemix.net/catalog/services/ibm-watson-machine-learning/) instance required for model storing and deployment (a free plan is offered).
- Create a [Cloud Object Storage Service](https://console.bluemix.net/catalog/infrastructure/cloud_object_storage/) instance required for storing scoring data (a free plan is offered).
- Create a [Apache Spark Service](https://console.bluemix.net/catalog/services/apache-spark/) instance required for batch scoring (a free plan is offered).
-  Make sure that you are using a Spark 2.0 kernel in notebook.

### Create the Telco Customer Churn Data Asset  

The Telco Customer Churn data is available on project's Github page.

1.  Download dataset from **Telco Customer Churn project** page [data folder](https://github.com/pmservice/wml-sample-models/tree/master/spark/customer-satisfaction-prediction/data) to local machine filesystem.
2.  After downloading the dataset to your local machine, read it into Spark DataFrame.

In [1]:
#@hidden_cell
import os

_filename = 'WA_Fn%20UseC_%20Telco%20Customer%20Churn.csv'
if os.path.isfile(_filename):
    os.remove(_filename)
del _filename

In [2]:
import wget

link_to_data = 'https://github.com/pmservice/wml-sample-models/raw/master/spark/customer-satisfaction-prediction/data/WA_Fn%20UseC_%20Telco%20Customer%20Churn.csv'
filename = wget.download(link_to_data)

print(filename)

WA_Fn%20UseC_%20Telco%20Customer%20Churn.csv


<a id="load"></a>
## 2. Load and explore data

In this section you will load the data as an Apache® Spark DataFrame and perform a basic exploration. 

Once you see the file on **Files** section, you can insert it to your notebook by selecting **Insert SparkSession DataFrame**

In [3]:
spark = SparkSession.builder.getOrCreate()

df_data = spark.read\
  .format('org.apache.spark.sql.execution.datasources.csv.CSVFileFormat')\
  .option('header', 'true')\
  .option('inferSchema', 'true')\
  .option('nanValue', ' ')\
  .option('nullValue', ' ')\
  .load(filename)

Explore the loaded data by using the following Apache® Spark DataFrame methods:
-  print schema
-  count all records
-  show distribution of label classes

In [4]:
df_data.printSchema()

print("Number of fields: %3g" % len(df_data.schema))

root
 |-- customerID: string (nullable = true)
 |-- gender: string (nullable = true)
 |-- SeniorCitizen: integer (nullable = true)
 |-- Partner: string (nullable = true)
 |-- Dependents: string (nullable = true)
 |-- tenure: integer (nullable = true)
 |-- PhoneService: string (nullable = true)
 |-- MultipleLines: string (nullable = true)
 |-- InternetService: string (nullable = true)
 |-- OnlineSecurity: string (nullable = true)
 |-- OnlineBackup: string (nullable = true)
 |-- DeviceProtection: string (nullable = true)
 |-- TechSupport: string (nullable = true)
 |-- StreamingTV: string (nullable = true)
 |-- StreamingMovies: string (nullable = true)
 |-- Contract: string (nullable = true)
 |-- PaperlessBilling: string (nullable = true)
 |-- PaymentMethod: string (nullable = true)
 |-- MonthlyCharges: double (nullable = true)
 |-- TotalCharges: double (nullable = true)
 |-- Churn: string (nullable = true)

Number of fields:  21


As you can see, the data contains 21 fields. "Churn" field is the one we would like to predict (label).

In [10]:
print("Total number of records: " + str(df_data.count()))

Total number of records: 7043


Data set contains 7043 records.

Now you will check if all records have complete data.

In [11]:
df_complete = df_data.dropna()

print("Number of records with complete data: %3g" % df_complete.count())

Number of records with complete data: 7032


You can see that there are some missing values you can investigate that all missing values are present in `TotalCharges` feature. We will use dataset with missing values removed for model training and evaluation.

Now you will inspect distribution of classes in label column.

In [12]:
df_complete.groupBy('Churn').count().show()

+-----+-----+
|Churn|count|
+-----+-----+
|   No| 5163|
|  Yes| 1869|
+-----+-----+



<a id="model"></a>
## 3. Create an Apache® Spark machine learning model

In this section you will learn how to prepare data, create an Apache® Spark machine learning pipeline, and train a model.

### 3.1: Prepare data

In this subsection you will split your data into: train, test and predict datasets.

In [13]:
(train_data, test_data, predict_data) = df_complete.randomSplit([0.8, 0.18, 0.02], 24)

print("Number of records for training: " + str(train_data.count()))
print("Number of records for evaluation: " + str(test_data.count()))
print("Number of records for prediction: " + str(predict_data.count()))

Number of records for training: 5621
Number of records for evaluation: 1263
Number of records for prediction: 148


As you can see our data has been successfully split into three datasets: 

-  The train data set, which is the largest group, is used for training.
-  The test data set will be used for model evaluation and is used to test the assumptions of the model.
-  The predict data set will be used for prediction.

### 3.2: Create pipeline and train a model

In this section you will create an Apache® Spark machine learning pipeline and then train the model.

In the first step you need to import the Apache® Spark machine learning packages that will be needed in the subsequent steps.

In [14]:
from pyspark.ml.feature import StringIndexer, IndexToString, RFormula
from pyspark.ml.classification import LogisticRegression
from pyspark.ml.evaluation import MulticlassClassificationEvaluator
from pyspark.ml import Pipeline, Model

In the following step, convert all the predictors to features vector and label feature convert to numeric.

In [15]:
lab = StringIndexer(inputCol = 'Churn', outputCol = 'label')
features = RFormula(formula = "~ gender + SeniorCitizen +  Partner + Dependents + tenure + PhoneService + MultipleLines + InternetService + OnlineSecurity + OnlineBackup + DeviceProtection + TechSupport + StreamingTV + StreamingMovies + Contract + PaperlessBilling + PaymentMethod + MonthlyCharges + TotalCharges - 1")

Next, define estimators you want to use for classification. Logistic Regression is used in the following example.

In [16]:
lr = LogisticRegression(maxIter = 10)

Let's build the pipeline now. A pipeline consists of transformers and an estimator.

In [17]:
pipeline_lr = Pipeline(stages = [features, lab , lr])

Now, you can train your Logistic Regression model by using the previously defined **pipeline** and **train data**.

In [18]:
model_lr = pipeline_lr.fit(train_data)

You can check your **model accuracy** now. To evaluate the model, use **test data**.

In [19]:
predictions = model_lr.transform(test_data)
evaluator = MulticlassClassificationEvaluator(labelCol="label", predictionCol="prediction", metricName="accuracy")
accuracy = evaluator.evaluate(predictions)

print("Test dataset:")
print("Accuracy = %3.2f" % accuracy)

Test dataset:
Accuracy = 0.81


You can tune your model now to achieve better accuracy. For simplicity of this example tuning section is omitted.

<a id="persistence"></a>
## 4. Persist model

In this section you will learn how to store your pipeline and model in Watson Machine Learning repository by using python client libraries.

First, you must import Watson Machine Learning repository client libraries.

**Note**: Apache® Spark 2.0 is required.

In [20]:
from repository.mlrepositoryclient import MLRepositoryClient
from repository.mlrepositoryartifact import MLRepositoryArtifact
from repository.mlrepository import MetaProps

Authenticate to Watson Machine Learning service on Bluemix.

**Action**: Put credentials from your instance of Watson Machine Learning service in Bluemix here.</div>

In [21]:
wml_credentials = {
  "url": "https://ibm-watson-ml.mybluemix.net",
  "access_key": "***",
  "username": "***",
  "password": "***",
  "instance_id": "***"
}

In [22]:
ml_repository_client = MLRepositoryClient(wml_credentials['url'])
ml_repository_client.authorize(wml_credentials['username'], wml_credentials['password'])

Create model artifact (abstraction layer).

In [23]:
model_artifact = MLRepositoryArtifact(ml_artifact=model_lr, training_data=train_data, name="Customer Churn Model" )

**Tip**: The MLRepositoryArtifact method expects a trained model object, training data, and a model name. (It is this model name that is displayed by the Watson Machine Learning service).

### 4.1: Save pipeline and model

In this subsection you will learn how to save pipeline and model artifacts to your Watson Machine Learning instance.

In [24]:
saved_model = ml_repository_client.models.save(model_artifact)

Get saved model metadata from Watson Machine Learning.

**Tip**: Use *meta.available_props()* to get the list of available props.

In [25]:
print(saved_model.meta.available_props())

['inputDataSchema', 'evaluationMetrics', 'pipelineVersionHref', 'modelVersionHref', 'trainingDataRef', 'pipelineType', 'creationTime', 'lastUpdated', 'label', 'authorEmail', 'trainingDataSchema', 'authorName', 'version', 'modelType', 'runtime', 'evaluationMethod']


In [26]:
print("creationTime: " + str(saved_model.meta.prop("creationTime")))
print("modelType: " + saved_model.meta.prop("modelType"))
print("pipelineType: " + saved_model.meta.prop("pipelineType"))
print("runtime: " + str(saved_model.meta.prop("runtime")))
print("label: " + saved_model.meta.prop("label"))
print("modelVersionHref: " + saved_model.meta.prop("modelVersionHref"))

creationTime: 2017-10-27 09:54:10.341000+00:00
modelType: sparkml-model-2.0
pipelineType: sparkml-pipeline-2.0
runtime: spark-2.0
label: Churn
modelVersionHref: https://ibm-watson-ml.mybluemix.net/v2/artifacts/models/e1593da5-9917-4130-809f-04b2f8a00f6c/versions/73a24c42-2808-4af0-bfe8-2e874c6b8ced


**Tip**: **modelVersionHref** is our model unique indentifier in the Watson Machine Learning repository.

In [27]:
published_model_ID = saved_model.uid

print("Model Id: " + str(published_model_ID))

Model Id: e1593da5-9917-4130-809f-04b2f8a00f6c


**Model Id** can be used to retrive latest model version from Watson Machine Learning instance.

### 4.2: Load model

In this subsection you will learn how to load back saved model from specified instance of Watson Machine Learning.

In [28]:
loadedModelArtifact = ml_repository_client.models.get(saved_model.uid)

You can print for example model name to make sure that model has been loaded correctly.

In [29]:
print(str(loadedModelArtifact.name))

Customer Churn Model


As you can see the name is correct. You have already learned how save and load the model from Watson Machine Learning repository.

<a id="visualization"></a>
## 5. Predict locally and visualize

In this section you will learn how to load data from batch scoring and visualize the prediction results with plotly package.

### 5.1: Make local prediction using previously loaded model and test data

In this subsection you will score *predict_data* data set.

In [30]:
predictions = loadedModelArtifact.model_instance().transform(predict_data)

Preview the results by calling the *show()* method on the predictions DataFrame.

In [31]:
predictions.show(5)

+----------+------+-------------+-------+----------+------+------------+-------------+---------------+--------------+------------+----------------+-----------+-----------+---------------+--------------+----------------+--------------------+--------------+------------+-----+--------------------+-----+--------------------+--------------------+----------+
|customerID|gender|SeniorCitizen|Partner|Dependents|tenure|PhoneService|MultipleLines|InternetService|OnlineSecurity|OnlineBackup|DeviceProtection|TechSupport|StreamingTV|StreamingMovies|      Contract|PaperlessBilling|       PaymentMethod|MonthlyCharges|TotalCharges|Churn|            features|label|       rawPrediction|         probability|prediction|
+----------+------+-------------+-------+----------+------+------------+-------------+---------------+--------------+------------+----------------+-----------+-----------+---------------+--------------+----------------+--------------------+--------------+------------+-----+----------------

By tabulating a count, you can see the split between labels.

In [32]:
predictions.select("prediction").groupBy("prediction").count().show(truncate=False)

+----------+-----+
|prediction|count|
+----------+-----+
|0.0       |114  |
|1.0       |34   |
+----------+-----+



### 5.2: Sample visualization of data with Plotly package

In this subsection you will explore prediction results with Plotly, which is an online analytics and data visualization tool.

**Example**: First, you need to install required packages. You can do it by running the following code. Run it only one time.

    !pip install plotly --user 
    !pip install cufflinks --user

In [33]:
import sys
import pandas
import plotly.plotly as py
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
import cufflinks as cf
import plotly.graph_objs as go

init_notebook_mode(connected=True)
sys.path.append("".join([os.environ["HOME"]])) 

Plot a pie chart that shows the predicted churn split. !!! By most important features: Contract, MonthlyCharges !!! predicted vs observed churns.

In [34]:
cumulative_stats = predictions.groupby(['label']).count()
labels_data_plot = ['No','Yes']
values_data_plot = [cumulative_stats.select('count').collect()[x][0] for x in range(2)]

In [35]:
product_data = [go.Pie(
            labels=labels_data_plot,
            values=values_data_plot,
    )]

product_layout = go.Layout(
    title='Churn',
)

fig = go.Figure(data=product_data, layout=product_layout)
iplot(fig)

With this data set, you might want to do some analysis of Mean Monthly Charges per churn class

In [36]:
y_data_plot = [predictions.groupby(['label']).mean().collect()[x][3] for x in range(2)]

In [37]:
age_data = [go.Bar(
            y= y_data_plot,
            x= labels_data_plot
            
    )]

age_layout = go.Layout(
    title='Mean Monthly Charges per churn class',
    xaxis=dict(
        title = "Churn",
        showline=False,),
    yaxis=dict(
        title = "Mean Monthly Charges",
        ),
)

fig = go.Figure(data=age_data, layout=age_layout)
iplot(fig)

Based on bar plot you created, you might make the following conclusion: The mean monthly charges for churn customers is higher than non-churn customers as you may expect.

<a id="scoring"></a>
## 6. Deploy model and score in a Cloud

In this section you will learn how to create batch deployment and to score a new data record by using the Watson Machine Learning REST API. 
For more information about REST APIs, see the [Swagger Documentation](http://watson-ml-api.mybluemix.net/).

#### 6.1 Save data for prediction to Object Storage

**Example**: First, you need to install required packages. You can do it by running the following code. Run it only one time.

    !pip install python-swiftclient --user

First, you must import Object Storage client libraries.

In [5]:
import swiftclient.client as swiftclient
import urllib

<div>**Action**: Put credentials from Object Storage Service in Bluemix here.</div>

In [6]:
objectstorage_credentials = {
  "auth_url": "https://identity.open.softlayer.com",
  "project": "***",
  "projectId": "***",
  "region": "***",
  "userId": "***",
  "username": "***",
  "password": "***",
  "domainId": "***",
  "domainName": "***",
  "role": "***"
}

In [7]:
connection_userid = objectstorage_credentials.get("userId")
connection_projectid = objectstorage_credentials.get("projectId")
connection_password = objectstorage_credentials.get("password")
connection_region = objectstorage_credentials.get("region")
connection_authurl = objectstorage_credentials.get("auth_url")

Now you will create container in Object Storage and copy [dataset](https://raw.githubusercontent.com/pmservice/wml-sample-models/master/spark/customer-satisfaction-prediction/data/scoreInput.csv) for batch scoring from **Telco Customer Churn project**.

In [8]:
score_filename = "customerInput.csv"
container_name = "batchjob"

In [9]:
conn = swiftclient.Connection(key = connection_password, authurl = connection_authurl + "/v3", auth_version = '3', os_options = {"project_id": connection_projectid, "user_id": connection_userid, "region": connection_region})
object_storage_info = conn.get_account()
if(container_name not in [x.get('name') for x in object_storage_info[1]]):
    conn.put_container(container = container_name)
container_info = conn.get_container(container_name)
if(score_filename not in [x.get('name') for x in container_info[1]]):
    scoring_data_connection = urllib.urlopen('https://raw.githubusercontent.com/pmservice/wml-sample-models/master/spark/customer-satisfaction-prediction/data/scoreInput.csv')
    conn.put_object(container = container_name, obj = score_filename, contents = scoring_data_connection, content_type = 'text/plain')
    scoring_data_connection.close()
    print "File '%s' successfully uploaded to container '%s'" % (score_filename, container_name)
else:
    print "File '%s' already exists in container '%s'" % (score_filename, container_name)
conn.close()

File 'customerInput.csv' successfully uploaded to container 'batchjob'


### 6.2: Create model deployment

Now you can create a batch scoring endpoint. Execute the following sample code that uses the **published Model Id** value to create the scoring endpoint to the Bluemix repository.

#### Create access token to Watson Machine Learning Service

To work with the Watson Machine Leraning REST API you must generate an access token. To do that you can use the following sample code:

First, you must import standard python libraries.

In [27]:
import urllib3, requests, json

headers = urllib3.util.make_headers(basic_auth='{username}:{password}'.format(username=wml_credentials['username'], password=wml_credentials['password']))
url = '{url}/v3/identity/token'.format(url=wml_credentials['url'])
response = requests.get(url, headers=headers)
mltoken = json.loads(response.text).get('token')

### 6.1: Create batch deployment

#### Get published_models url from instance details

In [28]:
endpoint_instance = wml_credentials['url'] + "/v3/wml_instances/" + wml_credentials['instance_id']
header = {'Content-Type': 'application/json', 'Authorization': 'Bearer ' + mltoken}

response_get_instance = requests.get(endpoint_instance, headers=header)

print(response_get_instance)
print(json.dumps(response_get_instance.text, indent=2))

<Response [200]>
"{\"metadata\":{\"guid\":\"5211de34-6f96-43ef-94ce-801d701e69b5\",\"url\":\"https://ibm-watson-ml.mybluemix.net/v3/wml_instances/5211de34-6f96-43ef-94ce-801d701e69b5\",\"created_at\":\"2017-10-10T06:51:16.664Z\",\"modified_at\":\"2017-10-26T08:55:37.279Z\"},\"entity\":{\"source\":\"Bluemix\",\"published_models\":{\"url\":\"https://ibm-watson-ml.mybluemix.net/v3/wml_instances/5211de34-6f96-43ef-94ce-801d701e69b5/published_models\"},\"usage\":{\"expiration_date\":\"2017-11-01T00:00:00.000Z\",\"computation_time\":{\"limit\":18000,\"current\":0},\"model_count\":{\"limit\":200,\"current\":4},\"prediction_count\":{\"limit\":5000,\"current\":10},\"deployment_count\":{\"limit\":5,\"current\":1}},\"plan_id\":\"3f6acf43-ede8-413a-ac69-f8af3bb0cbfe\",\"status\":\"Active\",\"organization_guid\":\"b0e61605-a82e-4f03-9e9f-2767973c084d\",\"region\":\"us-south\",\"account\":{\"id\":\"f52968f3dbbe7b0b53e15743d45e5e90\",\"name\":\"Umit Cakmak's Account\",\"type\":\"TRIAL\"},\"owner\":{\

In [29]:
endpoint_published_models = json.loads(response_get_instance.text).get('entity').get('published_models').get('url')

print(endpoint_published_models)

https://ibm-watson-ml.mybluemix.net/v3/wml_instances/5211de34-6f96-43ef-94ce-801d701e69b5/published_models


Execute the following sample code that uses the published_models endpoint to get deployments url.

#### Get the list of published models

In [32]:
header = {'Content-Type': 'application/json', 'Authorization': 'Bearer ' + mltoken}
response_get = requests.get(endpoint_published_models, headers=header)

print(response_get)
print(json.dumps(response_get.text, indent=2))

<Response [200]>
"{\"count\":1,\"resources\":[{\"metadata\":{\"guid\":\"d29e197d-2d96-4c48-be7e-288060057ff9\",\"url\":\"https://ibm-watson-ml.mybluemix.net/v3/wml_instances/5211de34-6f96-43ef-94ce-801d701e69b5/published_models/d29e197d-2d96-4c48-be7e-288060057ff9\",\"created_at\":\"2017-10-26T08:55:37.238Z\",\"modified_at\":\"2017-10-26T08:55:37.461Z\"},\"entity\":{\"runtime_environment\":\"spark-2.0\",\"author\":{},\"name\":\"Hey there Customer Churn Model\",\"label_col\":\"Churn\",\"training_data_schema\":{\"fields\":[{\"metadata\":{},\"type\":\"string\",\"name\":\"customerID\",\"nullable\":true},{\"metadata\":{},\"type\":\"string\",\"name\":\"gender\",\"nullable\":true},{\"metadata\":{},\"type\":\"integer\",\"name\":\"SeniorCitizen\",\"nullable\":true},{\"metadata\":{},\"type\":\"string\",\"name\":\"Partner\",\"nullable\":true},{\"metadata\":{},\"type\":\"string\",\"name\":\"Dependents\",\"nullable\":true},{\"metadata\":{},\"type\":\"integer\",\"name\":\"tenure\",\"nullable\":true}

#### Get published model deployment url

In [33]:
[endpoint_deployments] = [x.get('entity').get('deployments').get('url') for x in json.loads(response_get.text).get('resources') if x.get('metadata').get('guid') == saved_model.uid]

print(endpoint_deployments)

https://ibm-watson-ml.mybluemix.net/v3/wml_instances/5211de34-6f96-43ef-94ce-801d701e69b5/published_models/d29e197d-2d96-4c48-be7e-288060057ff9/deployments


#### Create batch deployment for published model

First import standard python library.

In [34]:
import base64

<div>**Action**:  Put your credentials from Apache Spark Service in Bluemix here</div>

In [35]:
spark_credentials = {
  "tenant_id": "***",
  "tenant_id_full": "***",
  "cluster_master_url": "***",
  "tenant_secret": "***",
  "instance_id": "***",
  "plan": "***"
}

In [36]:
encoded_spark_credentials_header = base64.b64encode(json.dumps({"credentials": spark_credentials, "version": "2.0"}))
result_filename = "result.csv"

In [37]:
header_batch = {'Content-Type': 'application/json', 'Authorization': "Bearer " + mltoken, 'X-Spark-Service-Instance': encoded_spark_credentials_header}
payload_batch = {"type":"batch", "name": "Umits Customer Churn Prediction", "description": "Batch Deployment", "input": {"source": { "container": container_name, "filename": score_filename, "fileformat": "csv", "inferschema": 1, "firstlineheader": "true", "type": "bluemixobjectstorage"}, "connection": {"userid": connection_userid, "projectid": connection_projectid, "password": connection_password, "region": connection_region, "authurl": connection_authurl}}, "output": { "target": { "container": container_name, "filename": result_filename, "fileformat": "csv", "firstlineheader": "true", "type": "bluemixobjectstorage"}, "connection": {"userid": connection_userid, "projectid": connection_projectid, "password": connection_password, "region": connection_region, "authurl": connection_authurl}}}

response_batch = requests.post(endpoint_deployments, json=payload_batch, headers=header_batch)

print(response_batch)
print(json.loads(response_batch.text).get("entity").get("status"))

<Response [201]>
INITIALIZING


You can see that your job has been submited to Spark.

#### Monitor job execution

Here you can check status of your batch scoring. When status of Spark job is `FINNISHED` the results will be written to result.csv file in Object Storage.

In [76]:
header_monitor_job = {'Content-Type': 'application/json', 'Authorization': "Bearer " + mltoken, 'X-Spark-Service-Instance': encoded_spark_credentials_header}

response_monitor_job = requests.get(endpoint_deployments, headers=header_monitor_job)

print(response_monitor_job)
print(json.dumps(response_monitor_job.text, indent=2))

<Response [200]>
"{\"count\":1,\"resources\":[{\"metadata\":{\"guid\":\"9f0b7f69-bcb8-47dd-9bd1-6396cfb38987\",\"url\":\"https://ibm-watson-ml.mybluemix.net/v3/wml_instances/5211de34-6f96-43ef-94ce-801d701e69b5/published_models/d29e197d-2d96-4c48-be7e-288060057ff9/deployments/9f0b7f69-bcb8-47dd-9bd1-6396cfb38987\",\"created_at\":\"2017-10-26T08:58:55.802Z\",\"modified_at\":\"2017-10-26T08:59:16.179Z\"},\"entity\":{\"runtime_environment\":\"spark-2.0\",\"name\":\"Umits Customer Churn Prediction\",\"description\":\"Batch Deployment\",\"published_model\":{\"author\":{},\"name\":\"Hey there Customer Churn Model\",\"url\":\"https://ibm-watson-ml.mybluemix.net/v3/wml_instances/5211de34-6f96-43ef-94ce-801d701e69b5/published_models/d29e197d-2d96-4c48-be7e-288060057ff9\",\"guid\":\"d29e197d-2d96-4c48-be7e-288060057ff9\",\"created_at\":\"2017-10-26T08:58:55.786Z\"},\"model_type\":\"sparkml-model-2.0\",\"status\":\"COMPLETED\",\"output\":{\"connection\":{\"projectid\":\"a574e627c6f34c67996c28f65a

**Preview scored data**

In this subsection you will load scored data.

Get data with scores from Object Storage Service in Bluemix.

In [77]:
conn = swiftclient.Connection(key = connection_password, authurl = connection_authurl + "/v3", auth_version = '3', os_options = {"project_id": connection_projectid, "user_id": connection_userid, "region": connection_region})

In [78]:
from StringIO import StringIO

In [79]:
results = conn.get_object(container_name, result_filename)
results_pdf = pandas.read_csv(StringIO(results[1]))

Preview the results by calling the *show()* method on the predictions in Pandas data frame.

In [80]:
results_pdf

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  9237-HQITU  Female              0      No         No       2          Yes   
1  3638-WEABW  Female              0     Yes         No      58          Yes   
2  8665-UTDHZ    Male              0     Yes        Yes       1           No   
3  8773-HHUOZ  Female              0      No        Yes      17          Yes   
4  4080-IIARD  Female              0     Yes         No      13          Yes   
5  6575-SUVOI  Female              1     Yes         No      25          Yes   
6  7495-OOKFY  Female              1     Yes         No       8          Yes   
7  0731-EBJQB  Female              0     Yes        Yes      52          Yes   
8  1891-QRQSA    Male              1     Yes        Yes      64          Yes   
9  5919-TMRGD  Female              0      No        Yes       1          Yes   

      MultipleLines InternetService       OnlineSecurity    ...      \
0                No     Fiber optic                   No    ...       
1               Yes             DSL                   No    ...       
2  No phone service             DSL                   No    ...       
3                No             DSL                   No    ...       
4               Yes             DSL                  Yes    ...       
5               Yes             DSL                  Yes    ...       
6               Yes     Fiber optic                   No    ...       
7                No              No  No internet service    ...       
8               Yes     Fiber optic                  Yes    ...       
9                No     Fiber optic                   No    ...       

           TechSupport          StreamingTV      StreamingMovies  \
0                   No                   No                   No   
1                  Yes                   No                   No   
2                   No                   No                   No   
3                   No                  Yes                  Yes   
4                  Yes                  Yes                   No   
5                  Yes                  Yes                   No   
6                   No                   No                   No   
7  No internet service  No internet service  No internet service   
8                  Yes                  Yes                  Yes   
9                   No                  Yes                   No   

         Contract PaperlessBilling              PaymentMethod  \
0  Month-to-month              Yes           Electronic check   
1        Two year              Yes    Credit card (automatic)   
2  Month-to-month               No           Electronic check   
3  Month-to-month              Yes               Mailed check   
4  Month-to-month              Yes           Electronic check   
5  Month-to-month              Yes    Credit card (automatic)   
6  Month-to-month              Yes    Credit card (automatic)   
7        One year              Yes           Electronic check   
8        Two year              Yes  Bank transfer (automatic)   
9  Month-to-month              Yes           Electronic check   

                                            features  \
0  (30,[2,3,4,5,6,8,10,12,14,16,18,20,22,24,25,28...   
1  (30,[3,4,5,7,9,10,13,14,17,18,20,23,24,28,29],...   
2  (30,[0,4,9,10,13,14,16,18,20,22,25,28,29],[1.0...   
3  (30,[2,4,5,6,9,10,12,14,16,19,21,22,24,26,28,2...   
4  (30,[3,4,5,7,9,11,13,14,17,19,20,22,24,25,28,2...   
5  (30,[1,3,4,5,7,9,11,12,14,17,19,20,22,24,28,29...   
6  (30,[1,3,4,5,7,8,10,13,14,16,18,20,22,24,28,29...   
7  (30,[4,5,6,24,25,28,29],[52.0,1.0,1.0,1.0,1.0,...   
8  (30,[0,1,4,5,7,8,11,12,15,17,19,21,23,24,27,28...   
9  (30,[2,4,5,6,8,10,12,14,16,19,20,22,24,25,28,2...   

                                rawPrediction  \
0    [-0.6683985606091145,0.6683985606091145]   
1        [3.94407920167498,-3.94407920167498]   
2  [-0.09214192316229042,0.09214192316229042]   
3      [0.747504546477989,-0.747504546477989]   
4    [0.59179

You can see the predictions for all the records in pandas data frame.

<a id="summary"></a>
## 7. Summary and next steps     

 You successfully completed this notebook! You learned how to use Apache Spark machine learning as well as Watson Machine Learning for model creation and deployment. Check out our _[Online Documentation](https://console.ng.bluemix.net/docs/services/PredictiveModeling/index.html?pos=2)_ for more samples, tutorials, documentation, how-tos, and blog posts. 

### Authors

**Umit Cakmak**, Data Scientist in IBM with a track record of developing enterprise-level applications that substantially increases clients' ability to turn data into actionable knowledge.

Copyright © 2017 IBM. This notebook and its source code are released under the terms of the MIT License.